File: src/types/session.ts
export type SessionState = "idle" | "connecting" | "verifying-host" | "authenticating" | "opening-shell" | "connected" | "disconnected" | "error";

export type TerminalEvent =
  | { event: "output"; data: { bytes: number[] } }
  | { event: "state"; data: { state: SessionState } }
  | { event: "closed"; data: { reason: string | null } };

export type HostVerification = {
  attemptId: string;
  host: string;
  port: number;
  keyType: string;
  fingerprint: string;
  status: "unknown" | "trusted";
  routeScope?: string;
};

export type KnownHost = {
  routeScope: string;
  host: string;
  port: number;
  keyType: string;
  fingerprint: string;
  createdAt: string;
  updatedAt: string;
};

export type ConnectRequest = {
  verificationAttemptId: string;
  profileId: string;
  credential: CredentialInput;
  jumpPreparationId?: string;
  cols: number;
  rows: number;
};

export type PrepareJumpConnectionResponse = {
  preparationId: string;
  targetVerification: HostVerification;
  jumpCredentialSaved: boolean;
};

export type TestConnectionRequest = Omit<ConnectRequest, "cols" | "rows">;

export type CredentialInput =
  | { mode: "session-only"; secret: string }
  | { mode: "remember-securely"; secret: string }
  | { mode: "stored" };

export type CredentialKind = "password" | "key-passphrase";

export type CredentialStatus = {
  vaultInitialized: boolean;
  vaultUnlocked: boolean;
  hasCredential: boolean;
  platformUnlockSupported: boolean;
  platformUnlockAvailable: boolean;
  platformUnlockConfigured: boolean;
};

export type PrivateKeyImport = { keyId: string; name: string };

export type ConnectResponse = {
  sessionId: string;
  credentialSaved: boolean;
};

export type TestConnectionResponse = {
  credentialSaved: boolean;
};

export type SftpEntryKind = "directory" | "file" | "symlink" | "other";

export type SftpEntry = {
  name: string;
  path: string;
  kind: SftpEntryKind;
  size: number | null;
  modified: number | null;
  permissions: number | null;
};

export type SftpDirectory = {
  path: string;
  entries: SftpEntry[];
};

export type SftpMetadata = Omit<SftpEntry, "name">;

export type RemoteImagePreview = {
  path: string;
  name: string;
  mimeType: "image/png" | "image/jpeg" | "image/webp" | "image/gif";
  width: number;
  height: number;
  size: number;
  dataBase64: string;
};

export type RemoteTextPreview = {
  path: string;
  name: string;
  encoding: string;
  language: string;
  size: number;
  lineCount: number;
  content: string;
};

export type LocalFileSelection = { grantId: string; name: string; size: number };
export type UploadDirectoryHistoryEntry = { remoteDirectory: string; lastUploadedAtMs: number };
export type TransferDirection = "upload" | "download";
export type TransferState = "queued" | "running" | "completed" | "failed" | "cancelled";
export type TransferJob = {
  id: string;
  sessionId: string;
  direction: TransferDirection;
  name: string;
  remotePath: string;
  totalBytes: number;
  transferredBytes: number;
  state: TransferState;
  errorCode: string | null;
};
export type TransferEvent = { event: "updated"; data: { job: TransferJob } };

File: src-tauri/src/domain/session.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use zeroize::Zeroizing;

use super::{CredentialInput, HostVerification};

pub type SessionId = Uuid;

#[derive(Clone, Copy, Debug, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum SessionState {
    Connecting,
    VerifyingHost,
    Authenticating,
    OpeningShell,
    Connected,
    Disconnected,
    Error,
}

#[derive(Clone, Serialize)]
#[serde(
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    tag = "event",
    content = "data"
)]
pub enum TerminalEvent {
    Output { bytes: Vec<u8> },
    State { state: SessionState },
    Closed { reason: Option<String> },
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HostKeyInfo {
    pub key_type: String,
    pub fingerprint: String,
}

pub struct ConnectRequest {
    pub connection: SshConnectionRequest,
    pub cols: u32,
    pub rows: u32,
}

pub struct SshConnectionRequest {
    pub host: String,
    pub port: u16,
    pub username: String,
    pub authentication: SshAuthentication,
}

pub enum SshAuthentication {
    Password {
        password: Zeroizing<String>,
    },
    PrivateKeyFile {
        path: String,
        passphrase: Option<Zeroizing<String>>,
    },
    PrivateKeyData {
        content: Zeroizing<Vec<u8>>,
        passphrase: Option<Zeroizing<String>>,
    },
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectProfileRequest {
    pub verification_attempt_id: Uuid,
    pub profile_id: Uuid,
    pub credential: CredentialInput,
    #[serde(default)]
    pub jump_preparation_id: Option<Uuid>,
    pub cols: u32,
    pub rows: u32,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TestConnectionProfileRequest {
    pub verification_attempt_id: Uuid,
    pub profile_id: Uuid,
    pub credential: CredentialInput,
    #[serde(default)]
    pub jump_preparation_id: Option<Uuid>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareJumpConnectionRequest {
    pub target_profile_id: Uuid,
    pub jump_verification_attempt_id: Uuid,
    pub jump_credential: CredentialInput,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareJumpConnectionResponse {
    pub preparation_id: Uuid,
    pub target_verification: HostVerification,
    pub jump_credential_saved: bool,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelJumpConnectionRequest {
    pub preparation_id: Uuid,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectResponse {
    pub session_id: SessionId,
    pub credential_saved: bool,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TestConnectionResponse {
    pub credential_saved: bool,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteRequest {
    pub session_id: SessionId,
    pub data: Vec<u8>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResizeRequest {
    pub session_id: SessionId,
    pub cols: u32,
    pub rows: u32,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionRequest {
    pub session_id: SessionId,
}
